File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.cc
#include <iostream>
#include <queue>
#include <vector>

using namespace std;
/**
v1, v2 두 경로 반드시 지나기
1 -> v1 -> v2 -> n
1 -> v2 -> v1 -> n 
중 최소

무방향 그래프이므로
[v1 -> 1] + [v2 -> n] + [v1 -> v2]
[v2 -> 1] + [v1 -> n] + [v1 -> v2]

시간복잡도 O(ElogV)
2e5 * 10 = 2e6 가능
*/
int n, m; 
int a, b, c;
int v1, v2;

#define P pair<int, int>
const int MX = 802;
vector<P> vec[MX];
long dist[MX];
bool ch[MX];

void get_dist(int s){
    priority_queue<P, vector<P>, greater<P>> pq;
    
    for(int i = 0; i < MX; i++){
        dist[i] = 2e9;
        ch[i] = 0;
    }
    
    pq.push({0, s});
    dist[s] = 0;
    
    while(pq.size()){
        int cur = pq.top().second;
        int d= pq.top().first;
        pq.pop();
        
        if(ch[cur]) continue;
        ch[cur] = 1;
        
        for(int i = 0; i < vec[cur].size(); i++){
            int nxt = vec[cur][i].first;
            int nd = vec[cur][i].second + d;
            
            if(dist[nxt] > nd){
                dist[nxt] = nd;
                pq.push({nd, nxt});
            }
        }
    }
}

int main()
{   
    cin >> n >> m;
    
    for(int i = 0; i < m; i++){
        cin >> a >> b >> c;
        
        vec[a].push_back({b,c});
        vec[b].push_back({a,c});
    }
    
    cin >> v1 >> v2;
    
    get_dist(v1);
    long d1 = dist[1];
    long d2 = dist[n];
    long d3 = dist[v2];
    
    get_dist(v2);
    long d4 = dist[1];
    long d5 = dist[n];
    
    long res = min(d1 + d5 + d3, d2 + d4 + d3);
    
    if(res >= 2e9) cout << -1;
    else cout << res;
}

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.java
import java.io.*;
import java.util.*;

public class Main {
	static final int MX = 1000;

	static int N, M, X;
	static List<int[]>[] edges = new List[MX + 1];

	static int[] dist = new int[MX + 1];

	static boolean[] ch = new boolean[MX + 1];

	public static void getDist(int st) {
		PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[0] - o2[0]);

		for (int i = 1; i <= N; i++) {
			dist[i] = Integer.MAX_VALUE;
			ch[i] = false;
		}

		pq.add(new int[] { 0, st });
		dist[st] = 0;

		while (!pq.isEmpty()) {
			int cur = pq.peek()[1];
			int d = pq.peek()[0];
			pq.poll();

			if (ch[cur])
				continue;
			ch[cur] = true;

			for (int[] edge : edges[cur]) {
				int nxt = edge[1];
				int nd = edge[0] + d;
				if (nd < dist[nxt]) {
					dist[nxt] = nd;
					pq.add(new int[] { nd, nxt });
				}
			}
		}
	}

	public static void main(String[] args) throws Exception {
		// System.setIn(new FileInputStream("src/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		for (int i = 1; i <= N; i++) {
			edges[i] = new LinkedList<>();

		}

		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			int c = Integer.parseInt(st.nextToken());
			edges[a].add(new int[] { c, b });
			edges[b].add(new int[] { c, a });
		}
		st = new StringTokenizer(br.readLine());
		int a = Integer.parseInt(st.nextToken());
		int b = Integer.parseInt(st.nextToken());

		long d1 = 0; // 1 -> a -> b -> N
		long d2 = 0; // 1 -> b -> a -> N
		getDist(a);

		d1 += dist[1]; // d1 += dist[1] + dist[b] 는 오버플로우 위험
        d1 += dist[b];
		d2 += dist[b];
        d2 += dist[N];

		getDist(b);
		d1 += dist[N];
		d2 += dist[1];

		long mn = Math.min(d1, d2);
		if (d1 >= Integer.MAX_VALUE || d2 >= Integer.MAX_VALUE)
			mn = -1;

		System.out.println(mn);

	}
}